### ISSUE TITLE
Add Reinforced Concrete hashing algorithm to standard library

### ISSUE THREAD

**Issue by stefan2904** [Zokrates/ZoKrates#1244]
## Abstract

_Reinforced Concrete_ (https://www.rc-hash.info/) is a novel ZK-friendly hash function that (claims to be) "up to 15x faster than its predecessor Poseidon hash". I think this could be a great addition to the standard library.

Some more infos at the [ZK Hash Function Cryptanalysis Bounties](https://www.zkhashbounties.info/#h.2vxwizb5l1sd).

## Motivation

It is a novel and fast ZK-friendly hashing algorithm.

## Specification

There is a (open-access) paper at today's CCS:

> Lorenzo Grassi, Dmitry Khovratovich, Reinhard Lüftenegger, Christian Rechberger, Markus Schofnegger, and Roman Walch. 2022. Reinforced Concrete: A Fast Hash Function for Verifiable Computation. In Proceedings of the 2022 ACM SIGSAC Conference on Computer and Communications Security (CCS '22). Association for Computing Machinery, New York, NY, USA, 1323–1335. https://doi.org/10.1145/3548606.3560686

There is a reference implementation (in Rust): https://extgit.iaik.tugraz.at/krypto/zkfriendlyhashzoo

**Comment by Schaeff** [comment]
I looked a bit more into this and I found this in the paper:

> Restrictions and Future Work. Whereas RC clearly brings high
native and ZK performance, it also has its own restrictions. First
of all, a proof system should support lookup gates, as otherwise
the RC circuit would be quite big (we estimate it to be around 5000
constraints)

The schemes currently supported by ZoKrates do not support lookup gates. So this should probably come after we have support for lookups. It would still be interesting to implement it in ZoKrates, see how it performs, and use it as an example when we get to lookups.
